### PR TITLE
[improve][misc] Upgrade slf4j to 2.0.13

### PR DESCRIPTION
**Pull request by nodece**
### Motivation

Keep up-to-date, and this version supports [fluent logging api](https://www.slf4j.org/manual.html#fluent).

### Modifications

- Upgrade slf4j from 1.7.32 to 2.0.13
- Upgrade qpid-server to adapt the slf4j 2
- Use log4j-slf4j2-impl instead log4j-slf4j-impl
- Exclude the log4j-slf4j-impl and slf4j from some library

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

**Comment by lhotari** [comment]
@nodece license check fails

**Comment by nodece** [comment]
I downloaded the test image and then tried to test the CLI test, when proxy is start, `Failed to load class "org.slf4j.impl.StaticLoggerBinder"` doesn't be printed, but in the GHA, this message has been printed, please see https://productionresultssa13.blob.core.windows.net/actions-results/c3508721-b89b-426a-aa66-cd2a8c2f5a99/workflow-job-run-6b1ea721-31e7-5262-bdcc-795f87829cb5/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-04-04T17%3A59%3A14Z&sig=lMZ7Bk7gThEIkAHVFAbJLNchc2Nq2jgfkcFXlxcWf%2F8%3D&sp=r&spr=https&sr=b&st=2024-04-04T17%3A49%3A09Z&sv=2021-12-02
```
unzip -p pulsar-java-test-image.zst.zip | unzstd | docker load
cd $PULSAR_HOME
./build/run_integration_group.sh CLI
```

@lhotari Do you have any suggestions?

**Comment by lhotari** [comment]
> @lhotari Do you have any suggestions?

@nodece To debug, locally with java-test-image in addition what you provided, you'd have to do `export PULSAR_TEST_IMAGE_NAME=apachepulsar/java-test-image:latest` (ref: https://github.com/apache/pulsar/blob/d61f7585e8ee1d5e76ff1ff1eb0b3380bc7178c4/.github/workflows/pulsar-ci.yaml#L539) before running the tests locally. The default value for PULSAR_TEST_IMAGE_NAME is the `pulsar-test-latest-version` image (ref: https://github.com/apache/pulsar/blob/8957e353ded9ee24eccea349c7747da721d9e66a/tests/integration/src/test/java/org/apache/pulsar/tests/integration/containers/PulsarContainer.java#L54-L55).

**Comment by nodece** [comment]
Sorry, in my test, `PULSAR_TEST_IMAGE_NAME` has been set, but it still does not work.

**Comment by lhotari** [comment]
> Sorry, in my test, `PULSAR_TEST_IMAGE_NAME` has been set, but it still does not work.

@nodece What doesn't work?

**Comment by nodece** [comment]
> > Sorry, in my test, `PULSAR_TEST_IMAGE_NAME` has been set, but it still does not work.
>
> @nodece What doesn't work?

In the GHA, the CLI test always print the “Failed to load class "org.slf4j.impl.StaticLoggerBinder"”, but in my local, never print.

**Review by shoothzj — APPROVED** [review]

**Comment by nodece** [comment]
slf4j seems to have some issues, and once the issues are fixed, I will continue working.

File: pom.xml [hunk 1]
```diff
@@ -1192,7 +1200,12 @@ flexible messaging model and an intuitive client API.</description>
         <groupId>io.streamnative.oxia</groupId>
         <artifactId>oxia-client</artifactId>
         <version>${oxia.version}</version>
-        <classifier>shaded</classifier>
```
Comment (lhotari) [inline pom.xml:1195]: Is there a problem with the shaded oxia-client? Would it be worth reporting directly to https://github.com/streamnative/oxia ?

Comment (nodece) [reply pom.xml:1195]: Sure, this shaded includes the `org.slf4j`.

Comment (lhotari) [reply pom.xml:1195]: Isn't that something that shouldn't be shaded at all? would it be helpful to exclude it from shading?

Comment (lhotari) [reply pom.xml:1195]: sorry, the Java client is at https://github.com/streamnative/oxia-java

Comment (nodece) [reply pom.xml:1195]: Issue: https://github.com/streamnative/oxia-java/issues/129

I think the oxia-client should remove the slf4j from the shaded jar.

**Comment by BewareMyPower** [comment]
It's a breaking change that affects the down

In our KSN project (the private version of [KoP](https://github.com/streamnative/kop)), after upgrading to include this commit, logs cannot be printed anymore.

**Comment by BewareMyPower** [comment]
I tried adding the `reload4j` dependency and it didn't work.

```
[INFO] \- org.slf4j:slf4j-reload4j:jar:2.0.13:test
[INFO]    \- ch.qos.reload4j:reload4j:jar:1.2.22:test
```

```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
```

**Comment by nodece** [comment]
Could you provide the log4j/log4j2 and slf4j versions? 
I can help you solve the issues you are facing.

**Comment by mattisonchao** [comment]
@nodece 
Do we have any breaking changes for API/client API/client dependencies?  we should raise a PIP and discuss it to decide if we have a breaking change. :) 

cc @codelipenghui @BewareMyPower

**Comment by BewareMyPower** [comment]
We should document explicitly that users should exclude `slf4j-api` from any `pulsar-xxx` dependency. Otherwise, users might suffer from such simple dependency upgrading.

In addition, for existing users, we should note this breaking change in 3.3.0's release note.

**Comment by mattisonchao** [comment]
I mean we should discuss what strong reason pushed us to introduce this breaking change. :)

**Comment by BewareMyPower** [comment]
See my thread here: https://lists.apache.org/thread/4omvl62k4jhntj3rsywp14zzgh1l3l9q

For example, assuming users import the dependencies like:

```xml
<dependency>
  <groupId>org.apache.pulsar</groupId>
  <artifactId>pulsar-client-all</artifactId>
  <version>3.2.2</version>
</dependency>

<dependency>
  <groupId>org.slf4j</groupId>
  <artifactId>slf4j-simple</artifactId>
  <version>1.7.36</version>
</dependency>
```

After upgrading the `pulsar-client-all` dependency to 3.3.0, the `slf4j-simple:1.7.36` dependency won't work. So if users want to upgrade the `pulsar-client-all` dependency, they have to exclude the `slf4j-api` dependency from `pulsar-client-all` and import `slf4j-api` 1.7.x dependency explicitly.

**Comment by lhotari** [comment]
> We should document explicitly that users should exclude `slf4j-api` from any `pulsar-xxx` dependency. Otherwise, users might suffer from such simple dependency upgrading.
> 
> In addition, for existing users, we should note this breaking change in 3.3.0's release note.

I don't think that excluding is needed. Both maven and gradle use a dependency resolution strategy where the closest dependency wins when there are conflicts. The user would need to add slf4j-api dependency or use dependency management. Excluding isn't a recommended solution.

**Comment by lhotari** [comment]
> @nodece Do we have any breaking changes for API/client API/client dependencies? we should raise a PIP and discuss it to decide if we have a breaking change. :)
> 
> cc @codelipenghui @BewareMyPower

3.3.0 is for breaking changes like this. I explained in https://github.com/apache/pulsar/pull/22391#issuecomment-2095259400 why the user already has a problem in dependency management if this breaks their application.

The world is moving forward and also other projects are moving towards slf4j-api 2.0.x . That will happen sooner or later for all projects.

It's better to make this change in 3.3.0 than in the 4.0.0 LTS version so that we have time to adjust things if needed for 4.0.0 . Introducing major changes in 4.0.0 isn't the best possible solution.

File: pom.xml [hunk 2]
```diff
@@ -2701,5 +2713,13 @@ flexible messaging model and an intuitive client API.</description>
...
       </snapshots>
     </repository>
+    <!-- For the BDB JE dependency -->
+    <repository>
+      <id>oracle.releases</id>
+      <url>https://download.oracle.com/maven</url>
+      <snapshots>
+        <enabled>false</enabled>
+      </snapshots>
+    </repository>
```
Comment (lhotari) [inline pom.xml:2723]: It seems that this repository was accidentially added in this PR. It's being removed in #23132.

**Pull request merged**